### PR TITLE
[Transform] Capture the element types of the matmul in the matcher

### PR DESCRIPTION
**Pull request by qedawkins**
In the future when more than just F32 is supported, the element type will need to be captured to properly select padding values. This captures the element types in the matcher in preparation for that.

This patch is nearly NFC; some of the matcher cleanup also fixes boundary conditions for cases like `input(position, ...)` which previously checked `index <= 0`.

Also cleans up a test case in set_transform_strategy that wasn't properly testing the alignment condition for using the strategy.

### CODE REVIEW

**Review by nicolasvasilache — APPROVED** [review]

File: llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp [hunk 1]
```diff
@@ -465,6 +465,16 @@ static bool isProjectedMap(AffineMap map, int64_t projectedDim) {
   return true;
 }
 
+/// Helper to turn a potentially negative index to positive within the range [0,
+/// ub] and indicate whether the transformed index is in bounds.
```
Comment (ftynse) [inline llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp:469]: Ultra-nit: I think you'd want to specify that `ub` is not included in the range as either `[0, ub)` or somehow stating that explicitly.

File: llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp [hunk 2]
```diff
@@ -560,6 +573,27 @@ transform_ext::StructuredOpMatcher::input(int64_t position,
...
+                      << " capture element type");
+    int64_t updatedPosition = position;
+    if (!makeValidPositiveIndex(updatedPosition, linalgOp.getNumDpsInputs()))
+      return false;
+    auto shapedType = linalgOp.getDpsInputOperand(updatedPosition)
+                          ->get()
+                          .getType()
+                          .dyn_cast<ShapedType>();
+    if (!shapedType)
+      return false;
```
Comment (ftynse) [inline llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp:590]: Nit: it may be helpful to have a debug print here saying "not a shaped type" so we know when the matcher failed.

**Review by ftynse — COMMENTED** [review]
LGTM

File: llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp (hunk 1 above)
Comment (qedawkins) [reply llvm-external-projects/iree-dialects/lib/Transforms/TransformMatchers.cpp:469]: Oh no, I flipped the logic here once and forgot to update the comment. I opened #13546 to address the comments here. In the future I'll make sure to give more time for review, apologies this time though :(
